File: java/src/main/java/com/leetcode/OneFiveFour.java
package com.leetcode;

/**
 * @author: John
 * @date: 2020-01-30 17:56
 * @description: 154
 **/
public class OneFiveFour {
//    public static int findMin(int[] nums) {
//        int result = nums[0];
//        for (int i = 0; i < nums.length; i++) {
//            if (nums[i] < result) {
//                result = nums[i];
//            }
//        }
//        return result;
//    }

    public static int findMin(int[] nums) {
        int left = 0, right = nums.length - 1;
        while (left < right) {
            int mid = (left + right) / 2;
            if (nums[mid] > nums[right]) {
                left = mid + 1;
            } else if (nums[mid] < nums[right]) {
                right = mid;
            } else {
                right = right - 1;
            }
        }
        return nums[left];
    }
}

File: leetcodecpp/leetcode/OneFiveFour.cpp
//
// Created by John on 2020/1/30.
//

#include "OneFiveFour.h"
#include <vector>

using namespace std;

class Solution {
public:
    static int findMin(vector<int> &num) {
        int lo = 0;
        int hi = num.size() - 1;
        int mid = 0;
        while (lo < hi) {
            mid = lo + (hi - lo) / 2;
            if (num[mid] > num[hi]) {
                lo = mid + 1;
            } else if (num[mid] < num[hi]) {
                hi = mid;
            } else {
                hi--;
            }
        }
        return num[lo];
    }
};